### PR TITLE
8270836: Foolproof JavadocTester.checkOutput

### PR DESCRIPTION
**Pull request by pavelrappo**
When searching output for multiple strings it is easy to misexpress your intent, which could make a failing test pass. Such cases can be caught and reported to the programmer.

For example, for this search to succeed it is sufficient for the output to consist only of the `"error: bad"` string:

```
checkOutput(..., "error:", "error: bad", ...)
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270836](https://bugs.openjdk.java.net/browse/JDK-8270836): Foolproof JavadocTester.checkOutput

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to cf1e6a1edb9d5d28dbb8fba8596a343819dad6f9

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4811/head:pull/4811` \
`$ git checkout pull/4811`

Update a local copy of the PR: \
`$ git checkout pull/4811` \
`$ git pull https://git.openjdk.java.net/jdk pull/4811/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4811`

View PR using the GUI difftool: \
`$ git pr show -t 4811`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4811.diff">https://git.openjdk.java.net/jdk/pull/4811.diff</a>

</details>

### CODE REVIEW

File: test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java [hunk 1]
```diff
@@ -543,6 +557,26 @@ private void checkOutput(String path, String fileString, boolean expectedFound,
...
+    static boolean isFirstPrefixOfSecond(String a, String b) {
+        int i = 0;
+        int j = 0;
+        while (i < a.length() && j < b.length() && a.charAt(i) == b.charAt(j)) {
+            i++;
+            j++;
+        }
+        return i == a.length();
+    }
+
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java:579]: Why not `String.startsWith` ?

File: test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java [hunk 2]
```diff
@@ -543,6 +557,26 @@ private void checkOutput(String path, String fileString, boolean expectedFound,
...
 
+    private static boolean checkIfPrefixes(String... strings) {
+        String[] copy = Arrays.copyOf(strings, strings.length);
+        Arrays.sort(copy);
+        for (int i = 0; i < copy.length - 1; i++) {
+            if (isFirstPrefixOfSecond(copy[i], copy[i + 1]))
+                return true;
+        }
+        return false;
+    }
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java:568]: Interesting trade off in performance:
* either create sorted array and do `o(N)` scan 
* or just do `o(N^2)` scan

File: test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java (hunk 1 above)
Comment (pavelrappo) [reply test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java:579]: Good catch! I forgot that I need a qualitative answer not the quantitative (e.g. length of the common prefix).

File: test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java (hunk 2 above)
Comment (pavelrappo) [reply test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java:568]: For manually enumerated strings the difference between O(N*log(N)) and O(N^2) is all but absent due to a rather small N. That said, if you think that O(N^2) scan would read better, I can use it instead.

**Comment by pavelrappo** [comment]
What's interesting is that this PR caused 17 test failures in tier1, which is more than I had expected. I'll fix those use sites of `checkOutput` and update the PR. Looks like this PR has a good practical value.

File: test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java (hunk 2 above)
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java:568]: I don't have any data to support a strong feeling either way.  It did take a moment to realize what the code was doing.

File: test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java [hunk 3]
```diff
@@ -561,22 +561,12 @@ private static boolean checkIfPrefixes(String... strings) {
         String[] copy = Arrays.copyOf(strings, strings.length);
         Arrays.sort(copy);
         for (int i = 0; i < copy.length - 1; i++) {
-            if (isFirstPrefixOfSecond(copy[i], copy[i + 1]))
+            if (copy[i + 1].startsWith(copy[i]))
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java:564]: 👍

**Review by jonathan-gibbons — APPROVED** [review]

**Comment by hns** [comment]
Just wondering: wouldn't this also be a problem for non-prefix-matching strings? E.g.:

`checkOutput(..., "error: bad", "bad", ...)`

**Comment by pavelrappo** [comment]
> Just wondering: wouldn't this also be a problem for non-prefix-matching strings? E.g.:
> 
> `checkOutput(..., "error: bad", "bad", ...)`

Absolutely! It's not just prefixes and suffixes; any substrings would be problematic:

```
checkOutput(..., "error: bad message", "bad", ...)
```

Thanks for noticing this.

**Comment by pavelrappo** [comment]
The more test issues this patch uncovers, the more I wonder if we should change the `strings` parameter in `checkOutput` from vararg `String...` to `String`, which will force us to use a sequence of `checkOutput` calls. I think that if you search output for strings one string at a time, you are less likely to misexpress your intent. I also think that the vararg `String... strings` should be reserved for `checkOrder` where it makes perfect sense.

**Comment by jonathan-gibbons** [comment]
> The more test issues this patch uncovers, the more I wonder if we should change the `strings` parameter in `checkOutput` from vararg `String...` to `String`, which will force us to use a sequence of `checkOutput` calls. I think that if you search output for strings one string at a time, you are less likely to misexpress your intent. I also think that the vararg `String... strings` should be reserved for `checkOrder` where it makes perfect sense.

I disagree.     I don't see that a sequence of `checkOutput` calls is any more helpful or safe than a single varargs call.

**Comment by pavelrappo** [comment]
I've pushed one more commit, ee5c92f, to bring this PR to a safepoint and suspend the work. This is because of the unexpectedly big number of test issues that this PR has uncovered. I think that before resuming the work on this PR, we need to examine and fix each of the affected tests.

**Pull request closed without merging**